README.md

```diff
@@ -39,14 +39,37 @@ Here's the first example to get you started: a counter that can go up or down. Y
     <script type="module">
       import { h, app } from "https://unpkg.com/hyperapp"
 
+      const cloneObject = function (obj) {
+        return JSON.parse(JSON.stringify(obj))
```

Comment:
This far from realistic. No one writes this.

Reply:
Is this actually what happens though? aka, is the state actually getting cloned and returned new instead of mutated?

Reply:
I don't understand what you're asking.

Reply:
You said "no one writes like this"

Nobody writes cloneObject functions? Your library requires the `state` object be cloned because mutating the variable will not work/cause the changes to be flushed to the UI.

Reply:
@brandonros  Writing a cloneObject function is a very inefficient way to update the state, as you will have to process everything in a potentially large and deep state tree in order to alter a single value. Every single time a value changes (even in, say, an animation at 60fps...)

The way Jorge demonstrated is actually _does_ make the top level a new instance, and thus triggers a rerender. If your state is

```
{ foo: {flip: 1, flap: 2, flop: 3}, bar: {...}, baz: {...}}
```

and you say: 

```
{
  ...state,
  foo: {
    ...state.foo,
    flap: 42
  }
}
```

... then the state object itself will be new. `state.foo` will be new. `state.foo.flap` will have a new value. _But_ `state.bar` is the _same_ as before, and so is `state.foo.flip` (whatever it was).

Using the spread operator to compose new state not only saves you the trouble of cloning stuff -- because `bar` is still the same instance, it allows you to optimize rendering (basically skip diffing entire blocks of the vdom) any vnodes which depend on bar. 

That is the reason Jorges implementation of the `Increment` function is "official best practice".  It's unfortunate if people coming from other languages find it weird, but it is how it is done in all the official and unofficial examples I have seen anywhere. It's just something a newcomer to the framework will have to learn.

Reply:
@zaceno Let me try my example without the cloneObject function. I'm *pretty sure* it doesn't work, which makes me wonder... does the `spread` operator actually do some kind of shallow clone behind the scenes?

Reply:
yeah, take a look at this:

```
const decrement = function (event, state) {
  state.counter -= 1
  return state
}

const increment = function (event, state) {
  state.counter += 1
  return state
}

app({
  init: {
    counter: 0
  },
  view: function (state) {
    return h("main", {}, [
      h("h1", {}, `Value: ${state.counter}`),
      h("button", { onClick: function (event) { return decrement(event, state) } }, "-"),
      h("button", { onClick: function (event) { return increment(event, state) } }, "+")
    ])
  },
  node: document.getElementById("app")
})
```

this does not update the view on change

Reply:
I got it. Mutation is gross and causing weird things to subtly happen.

```
const decrement = function (event, state) {
  return Object.assign({}, state, {
    counter: state.counter - 1
  })
}

const increment = function (event, state) {
  return Object.assign({}, state, {
    counter: state.counter + 1
  })
}

app({
  init: {
    counter: 0
  },
  view: function (state) {
    return h("main", {}, [
      h("h1", {}, `Value: ${state.counter}`),
      h("button", { onClick: function (event) { return decrement(event, state) } }, "-"),
      h("button", { onClick: function (event) { return increment(event, state) } }, "+")
    ])
  },
  node: document.getElementById("app")
})
```

This is a better example :)

Reply:
You got it! Spread operator is just a shorter, more idiomatic `Object.assign({}, ...)` 👍

Reply:
Thank you, @zaceno!